src/cluster_legacy.c

```diff
@@ -1252,6 +1252,57 @@ void clusterInitLast(void) {
 
 /* Called when a cluster node receives SHUTDOWN. */
 void clusterHandleServerShutdown(void) {
+    if (nodeIsPrimary(myself) && server.auto_failover_on_shutdown) {
+        /* Find the first best replica, that is, the replica with the largest offset. */
+        int old_replica = 0;
+        client *best_replica = NULL;
+        listIter replicas_iter;
+        listNode *replicas_list_node;
+        listRewind(server.replicas, &replicas_iter);
+        while ((replicas_list_node = listNext(&replicas_iter)) != NULL) {
+            client *replica = listNodeValue(replicas_list_node);
+            /* This is done only when the replica offset is caught up, to avoid data loss.
+             * And 0x80100 is 8.1.0, we only support this feature in this version. */
+            if (replica->repl_data->replica_version < 0x80100) {
+                old_replica = 1;
+                best_replica = NULL;
+                break;
+            }
+            if (replica->repl_data->repl_state == REPLICA_STATE_ONLINE &&
+                replica->name && sdslen(replica->name->ptr) == CLUSTER_NAMELEN &&
+                replica->repl_data->repl_ack_off == server.primary_repl_offset) {
+                best_replica = replica;
+            }
+        }
+
+        if (best_replica) {
+            /* Send the CLUSTER FAILOVER FORCE REPLICAID node-id to all replicas since
+             * it is a shared replication buffer, but only the replica with the matching
+             * node-id will execute it. The caller will call flushReplicasOutputBuffers,
+             * so in here it is a best effort. */
+            char buf[128];
+            size_t buflen = snprintf(buf, sizeof(buf),
+                                     "*5\r\n$7\r\nCLUSTER\r\n"
+                                     "$8\r\nFAILOVER\r\n"
+                                     "$5\r\nFORCE\r\n"
+                                     "$9\r\nREPLICAID\r\n"
+                                     "$%d\r\n%s\r\n",
+                                     CLUSTER_NAMELEN,
+                                     (char *)best_replica->name->ptr);
+            /* Must install write handler for all replicas first before feeding
+             * replication stream. */
+            prepareReplicasToWrite();
+            feedReplicationBuffer(buf, buflen);
+            serverLog(LL_NOTICE, "Perform auto failover to replica %s on shutdown.", (char *)best_replica->name->ptr);
+        } else {
```

Comment:
I would suggest testing the negative cases first and return immediately, generally speaking. this would help reduce code indentation.